### ISSUE TITLE
align applets with relative position

### ISSUE THREAD

**Issue by schmittm** [mate-desktop/mate-panel#62]
Fix the somewhat "brokenness" of mate-panel (which led to the feature-trimming redesign of gnome-panel) in a way that the freedom for the user to place the applets and buttons wherever he likes is retained.

Problem specification:
When one resizes the screen resolution (and possible when other race conditions occur), the items on the panel tend to be shuffled around and be misaligned.

Possible solution: 
Position of items should be remembered in a smart and _relative_ way, regardless of the size of the panel.  That means if one puts an item on the very right side, it should _always_ be there, at the very end of the panel. All items placed directly side by side without spacing between them, should _always_ retain their order and null-spacing relative to the others. If the panel gets smaller, the items order should _never_ be touched, the space between the items should be adjusted instead. It should be tried to retain the relative position as good as possible under all circumstances.
If the size gets to small to place every item on the panel, a drop-down-mechanism or hide-and-show feature of groups of items depending on where one puts the mouse-cursor over, could be introduced. Maybe even some sort or user-setable prioritization of what items should _always_ be shown would be nice. For example the mate-menu is one that should be shown under all circumstances, it should default to that setting, but a user should be allowed to change that nevertheless. The systray (notification area) on the other hand, should default to collapse to one general icon if space on the panel gets scarce. User-added items should default to a higher-than-normal but lower-than-highest priority.
Bottom line, smart, sane defaults, user-configurable.

What do you think?

regards
Michael
##

**Comment by milankni** [comment]
+1.
Similar situation occurs when starting the same desktop locally (2560 px wide) and remotely (1100 px wide). I assume that every notebook user who ever connects external display faces similar issue.

Even there there is more than enough space on the panel after MATE desktop starts on the smaller screen size, the applets on the right side (typically notification area, clock, network speed indicator, logoff button) get once in a while re-sorted.

Once I log locally again (wide screen), the mentioned applets remain scattered across centre to the right part with various size of space between them.

Use of "Lock" feature does not have any impact on the behaviour.

**Comment by opnoack** [comment]
+1 But personally I don't see this as a "feature request". It's just a plain bug and it's over three years old.

**Comment by draeath** [comment]
This remains an issue. It is especially irritating in the context of virtual machines, as the display geometry is far more fluid in such an environment. It seems any time I resize the VM window, I end up with, essentially, all my panel items randomly positioned.

**Comment by mvglasow** [comment]
+1 Just had a similar issue for Ubuntu MATE rejected with the request to file it upstream.

Internally, as apparent in the config, the placement of a widget on the panel can be relative to its left or right edge (a negative X value will align it with the right edge).

However, when moving widgets, their position invariably gets stored as offset from left, which becomes apparent if the panel is resized and is a major nuisance when cleaning up a messed-up layout, as I'd have to manually edit the configuration in order to arrange my panel widgets the way they are supposed to be.

I therefore suggest introducing some logic to decide whether to store a widget's position relative to the left or to the right edge. Decision criteria could be:

* What is the alignment of the closest widget in either direction? (An edge would be treated like a widget aligned to that edge.) Use that alignment if it is unambiguous (i.e. unless there are widgets at either side at an equal distance with different alignments).
* Failing that, determine if the widget is closer to the left or to the right edge, then align it to that edge.
* If that still is ambiguous, stick with the current alignment of the widget.

**Comment by C0bb3r** [comment]
I made a similar request in 2014 (https://github.com/mate-desktop/mate-panel/issues/183). I didn't see this one when I made it. It is obvious that it is a problem.
